[Enhancement] Add be jvm memory metrics

nillin0 opened the pull request:
## Why I'm doing:

Some JNI calls are used to interact with Java in the BE.

## What I'm doing:

Adding JVM metrics to the BE, collecting various memory usage statistics during JVM runtime.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 4.0
  - [ ] 3.5
  - [ ] 3.4
  - [ ] 3.3

nillin0 committed:
Add be jvm memory metrics

Signed-off-by: wenbinlin <wenbinlin@tencent.com>

kevincai commented on be/src/util/jvm_metrics.cpp:
```diff
@@ -0,0 +1,233 @@
+void JVMMetrics::install(MetricRegistry* registry) {
+    if (!registry->register_hook("jvm_metrics", [this] { 
+            if (!update().ok()) {
+                LOG(WARNING) << "update jvm metrics failed: " << update().to_string();
```
doesn't look right to call update() repeatedly here

nillin0 committed:
fix update call repeatedly and remove redundant semicolon

Signed-off-by: wenbinlin <wenbinlin@tencent.com>

nillin0 committed:
fix and use clang-format

disable jvm_metrics by default
replace DCHECK with CHECK_JNI_EXCEPTION to handle nullptr returns
use clang-format

Signed-off-by: wenbinlin <wenbinlin@tencent.com>

nillin0 commented on be/src/common/config.h:
```diff
@@ -568,6 +568,8 @@ CONF_mBool(enable_token_check, "true");
 // to open/close system metrics
 CONF_Bool(enable_system_metrics, "true");
 
+CONF_Bool(enable_jvm_metrics, "true");
```
Done.

nillin0 commented on be/src/util/jvm_metrics.cpp:
```diff
@@ -0,0 +1,234 @@
+Status JVMMetrics::init() {
+    RETURN_IF_ERROR(detect_java_runtime());
+
+    JNIEnv* env = JVMFunctionHelper::getInstance().getEnv();
```
env had been checked in JVMFunctionHelper::getInstance().
```cpp
JVMFunctionHelper& JVMFunctionHelper::getInstance() {
    if (_env == nullptr) {
        _env = getJNIEnv();
        CHECK(_env != nullptr) << "couldn't got a JNIEnv";
    }
    static JVMFunctionHelper helper;
    return helper;
}
```

kevincai replied:
check will crash the process in release mode. shouldn't be implemented this way.

nillin0 committed:
check JNIEnv when initialize

Signed-off-by: wenbinlin <wenbinlin@tencent.com>

kevincai commented:
@nillin0 there is a code conflict with the current main branch, can you rebase the code to solve the conflict.

nillin0 committed:
add jvm metrics test

Signed-off-by: wenbinlin <wenbinlin@tencent.com>

kevincai commented on be/src/tools/mock_jvm.c:
```diff
@@ -30,8 +30,7 @@ _JNI_IMPORT_OR_EXPORT_ jint JNICALL JNI_CreateJavaVM(JavaVM** pvm, void** penv,
     fprintf(stderr, "unsupported call jni function, JAVA_HOME is required\n ");
-    exit(1);
-    return 1;
+    return JNI_ERR;
```
this mock_jvm.so is not supposed to be dynamically linked during the runtime, it is just a mock to workaround the binary link issue. Shall cry loudly if it is wrongly loaded in runtime.

kevincai commented on be/src/util/jvm_metrics.cpp:
```diff
@@ -0,0 +1,243 @@
+#include "util/jvm_metrics.h"
```
Please add license header.

kevincai commented on be/src/util/jvm_metrics.h:
```diff
@@ -0,0 +1,85 @@
+#pragma once
```
add license header for new files.

nillin0 committed:
add license and restore mock_jvm.c

Signed-off-by: wenbinlin <wenbinlin@tencent.com>

wyb commented on be/src/util/jvm_metrics.cpp:
```diff
@@ -0,0 +1,257 @@
+
+    REGISTER_JVM_METRIC(heap_used);
+    REGISTER_JVM_METRIC(heap_committed);
+    REGISTER_JVM_METRIC(heap_max);
```
why not use label like FE JVM metrics?

```
jvm_heap_size_bytes{type="max"} 1024
jvm_heap_size_bytes{type="committed"} 1024
jvm_heap_size_bytes{type="used"} 1024
```

nillin0 replied:
have modified

nillin0 committed:
trans label like FE JVM metrics

Signed-off-by: wenbinlin <wenbinlin@tencent.com>
